pkg/doc/jsonld/document_loader.go

```diff
@@ -7,67 +7,211 @@ SPDX-License-Identifier: Apache-2.0
+	options := &documentLoaderOpts{
+		contextDBName: DefaultContextDBName,
+		contextFS:     EmbedFS,
+		documents:     EmbedContexts,
 	}
 
-	return cdl
+	for i := range opts {
+		opts[i](options)
+	}
+
+	store, err := storageProvider.OpenStore(options.contextDBName)
+	if err != nil {
+		return nil, fmt.Errorf("new document loader: %w", err)
+	}
+
+	if len(options.documents) > 0 { // preload documents into the underlying storage
+		if e := save(store, options.documents, options.contextFS); e != nil {
+			return nil, fmt.Errorf("save context documents: %w", e)
+		}
+	}
+
+	return &DocumentLoader{
+		store:                store,
+		remoteDocumentLoader: options.remoteDocumentLoader,
+	}, nil
 }
 
-// LoadDocument returns a RemoteDocument containing the contents of the JSON resource from the given URL (u).
-func (cdl *CachingDocumentLoader) LoadDocument(u string) (*ld.RemoteDocument, error) {
-	cdl.rwMutex.RLock()
-	if doc, cached := cdl.cache[u]; cached {
-		defer cdl.rwMutex.RUnlock()
+func save(store storage.Store, docs []ContextDocument, sys fs.FS) error {
+	var ops []storage.Operation
 
-		if cachedDoc, ok := doc.(*ld.RemoteDocument); ok {
-			return cachedDoc, nil
+	for _, doc := range docs {
+		rd := ld.RemoteDocument{
+			DocumentURL: doc.DocumentURL,
 		}
 
-		return nil, errors.New("invalid document entry")
+		content := doc.Content
+
+		if content == nil {
+			c, err := loadFromFS(doc.Path, sys)
+			if err != nil {
+				return fmt.Errorf("load from FS: %w", err)
+			}
+
+			content = c
+		}
+
+		d, err := ld.DocumentFromReader(bytes.NewReader(content))
+		if err != nil {
+			return fmt.Errorf("document from reader: %w", err)
+		}
+
+		rd.Document = d
+
+		b, err := json.Marshal(rd)
+		if err != nil {
+			return fmt.Errorf("marshal remote document: %w", err)
+		}
+
+		ops = append(ops, storage.Operation{Key: doc.URL, Value: b})
+	}
+
+	// TODO: Support new/updated contexts on fresh storage and on the one with existing data (migrations).
+	if err := store.Batch(ops); err != nil {
+		return fmt.Errorf("store batch of contexts: %w", err)
 	}
 
-	cdl.rwMutex.RUnlock()
+	return nil
+}
 
-	docFromLoader, err := cdl.nextLoader.LoadDocument(u)
+func loadFromFS(path string, sys fs.FS) ([]byte, error) {
+	f, err := sys.Open(path)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("open file: %w", err)
 	}
 
-	cdl.rwMutex.Lock()
-	if _, cached := cdl.cache[u]; !cached {
-		cdl.cache[u] = docFromLoader
+	defer func() {
+		if e := f.Close(); e != nil {
+			logger.Errorf("close file: %w", e)
+		}
+	}()
+
+	return io.ReadAll(f)
+}
+
+// LoadDocument resolves JSON-LD context document by document URL (u) either from storage or from remote URL.
+// If document is not found in the storage and remote DocumentLoader is not specified, ErrContextNotFound is returned.
+func (l *DocumentLoader) LoadDocument(u string) (*ld.RemoteDocument, error) {
```

Comment:
Do we still have a cache?

Reply:
yes, but not directly - caching is going to be supported by using appropriate storage provider, e.g.
```
jsonld.NewDocumentLoader(cachedstore.NewProvider(storageProvider, mem.NewProvider()))
```
here:
* storageProvider is the main storage provider
* mem.NewProvider() - implementation of cacheProvider (just an example, mem is not the proper option for the real usage)

I am going to update document loader used in bdd tests in this PR to demonstrate the idea. And add caching support for `aries-js-worker`, `aries-agent-mobile` and `aries-agent-rest` in a separate follow-up ticket.

cc @troyronda @Baha-sk